modules/core/client/components/StepNavigation.js

```diff
@@ -0,0 +1,172 @@
+import classnames from 'classnames';
+import Tooltip from './Tooltip';
+
+/**
+ * StepNavigation can contain three different buttons: Back, Next, Submit we can go through.
+ * Each of them has a related property onBack, onNext, onSubmit.
+ * See it at `/offer/host` or `/profile/{username}/references/new`.
+ * This component doesn't keep its own state.
+ *
+ * @param {number} currentStep - index of current step
+ *                               first step doesn't have Back button
+ *                               last step has Finish/Submit button instead of Next
+ * @param {number} numberOfSteps - amount of steps in total
+ * @param {boolean} [disabled] - disable Next button
+ * @param {string} [disabledReason] - message to show in tooltip of disabled button
+ * @param {function} onBack - method executed on clicking Back
+ * @param {function} onNext - method executed on clicking Next
+ * @param {function} onSubmit - method executed on clicking Submit
+ */
+
+const BackButton = ({ small, ...props }) => {
+  const { t } = useTranslation('core');
+  return (
+    <button
+      type="button"
+      className={classnames({
+        btn: true,
+        'btn-lg': small,
+        'btn-primary': small,
+        'btn-action': !small,
+        'btn-link': !small,
+      })}
+      aria-label={t('Previous section')}
+      {...props}
+    >
+      <span className="icon-left" aria-hidden="true"></span>
+      {t('Back')}
+    </button>
+  );
+};
+BackButton.propTypes = { small: PropTypes.bool };
+
+const NextButton = ({ small, ...props }) => {
+  const { t } = useTranslation('core');
+
+  return (
+    <button
+      type="button"
+      className={classnames({
+        btn: true,
+        'btn-lg': small,
+        'btn-action': !small,
+        'btn-primary': true,
+      })}
+      aria-label={t('Next section')}
+      {...props}
+    >
+      {t('Next')}
+      {small && <span className="icon-right" aria-hidden="true"></span>}
+    </button>
+  );
+};
+NextButton.propTypes = { small: PropTypes.bool };
+
+const SubmitButton = ({ small, ...props }) => {
+  const { t } = useTranslation('core');
+  return (
+    <button
+      type="submit"
+      className={classnames({
+        btn: true,
+        'btn-lg': small,
+        'btn-action': !small,
+        'btn-primary': true,
+      })}
+      aria-label={t('Finish editing and save')}
+      {...props}
+    >
+      {t('Finish')}
+      {small && <span className="icon-ok" aria-hidden="true"></span>}
+    </button>
+  );
+};
+SubmitButton.propTypes = { small: PropTypes.bool };
+
+export default function StepNavigation({
+  currentStep,
+  numberOfSteps,
+  disabled = false,
+  disabledReason = '',
+  onBack,
+  onNext,
+  onSubmit,
+}) {
+  /**
+   * We'll reuse the buttons and tooltip defined below
+   * either with different children or with additional props.
+   * We use React.cloneElement(element, props, children) for that purpose:
+   * https://reactjs.org/docs/react-api.html#cloneelement
```

Comment:
Is the `cloneElement` really needed? Caused me to look twice, wondering what it is for.

My guess would be the intention is to improve performance by not re-creating buttons all the time, but I don't think this would be very heavy, as the button components are not complex.

Additionally, the "source" buttons (to be cloned) are being created on each render anyway (as it's a function component), so I think it is not saving much.

Or perhaps there is a reason I'm missing?

Reply:
The only reason was that the code looked less repetitive to me. With `cloneElement`, I was able to pass the props and children to buttons and Tooltip once, and then only tweak a little bit.
The alternative was to assign the props to variables and pass them to buttons and tooltip each time, which felt more repetitive, but maybe it's clearer?
Or maybe there is an alternative, like creating the navigation only once and styling it with media queries.
Does this make sense?

Reply:
changed in b8238c24c77407e89f24ab2aff5d25a9c4996c3b

it's clearer without cloneElement